File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  @non_alphanumeric_or_dashes ~r([^[:alnum:]\-])u

  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> convert_to_words()
    |> Enum.reduce(%{}, &increment_tally_for_word/2)
  end

  defp convert_to_words(sentence) do
    sentence
    |> String.downcase()
    |> String.split(@non_alphanumeric_or_dashes, trim: true)
  end

  defp increment_tally_for_word(word, acc) do
    Map.update(acc, word, 1, &(&1 + 1))
  end
end

File: javascript/word-count/word-count.js
const WORD = /\b[\w']+\b/g

export const countWords = sentence => {
  const words = convertToWords(sentence)
  return words.reduce(incrementTallyForWord, {})
}

function convertToWords(sentence) {
  return sentence.toLowerCase().match(WORD)
}

function incrementTallyForWord(acc, word) {
  acc[word] ? acc[word] += 1 : acc[word] = 1
  return acc
}
